Agasc1p7

malgosias opened the pull request:
This is a pull request for updated create*.py scripts changed so that they work with AGASC V1.7, and V1.7 test script which:
- confirms that all the fields in all the columns match values in the input agasc1p7.h5 file (new h5 file created from Jean's fits files is in /home/malgosia/git/agasc/agasc1p7.h5, branch agasc1p7)
- confirms that there are no differences between agasc1p7.h5 file and 1.6 file in columns other than new mag_aca, mag_aca_err and rsv 1-3 columns.

These tests fail for one star entry and it is not clear to me why: the AGASC_ID is OK, but RA and DEC are slightly different, there is also a difference in COLOR2 column (and I did not check the remaining columns for this one failing entry).

What is missing is a confirmation that we do see expected differences between agasc1p7.h5 file and 1.6 in new mag_aca, mag_aca_err and rsvd 1-3 columns. But I would like to first understand the origin of this one entry that fails the above tests. Let me know if you can think of any reasons for this.

jeanconn commented:
What's the agasc id on the fail?

jeanconn commented:
It looks to me like the h5 file built from the fits file has the problem that there are two entries for star 154534513.  I don't see where this is a mistake in the code that makes the h5 files so I suspect this is a bug in AGASC 1.6 (will confirm) and we'll need to decide if AGASC 1.7 should fix the bug.

jeanconn commented:
I've confirmed that we have two stars that share this AGASC_ID AGASC 1.6.  It looks like you can basically crash SAUSAGE on this if you manually select the star (or stars) with the shared id, but I consider this a no-impact bug and think we should just open an issue to fix this for the next AGASC.  Is there anything else we should confirm on that course of action @taldcroft ?

taldcroft commented on test_agasc1p7.py:
```diff
@@ -0,0 +1,47 @@
+id_map2 = dict(zip(agasc_ids2, idx))
+id_map3 = dict(zip(agasc_ids3, idx))
+
+h5inp_id_map = [id_map2[agasc_id] for agasc_id in agasc_ids]
```
Why is this ID map required?  The files should have exactly the same number of rows, in exactly the same order, right?

malgosias replied:
For some reason they don't have the same order. I'll double check this but without the ID map I was getting Assertion Errors.

taldcroft commented on test_agasc1p7.py:
```diff
@@ -0,0 +1,47 @@
+
```
Add module docstring saying what this script is doing and why.

taldcroft commented on test_agasc1p7.py:
```diff
@@ -0,0 +1,47 @@
+import numpy as np
+import tables
+
+h5 = tables.openFile('agasc1p7.h5', mode='r')
```
This file should be called (here) something like `agasc1p7_from_fits.h5`, and the other one should be `agasc1p7_from_jupyter.h5`.  Otherwise the heritage is very confusing.  You can do this with soft links and add a code comment describing where each file came from (first from the `create_agasc_h5.py` and second from the Jupyter notebook).

These docs won't be wasted effort since you will need to describe the file creation process in detail for the final AGASC 1.7 README (or wherever we decide that needs to go).

taldcroft commented on create_mini_agasc_h5.py:
```diff
@@ -5,12 +5,12 @@
 
 
-example_file = '/data/agasc1p6/agasc/n0000/0001.fit'
+example_file = '/proj/sot/ska/data/agasc1p7/agasc/n0000/0001.fit'
```
For this script you should use the version I created earlier that has a command line input to create either 1p6 or 1p7.

taldcroft replied:
I'm talking about the entire create_mini_agasc_h5.py script, not the `example_file`.  See: https://github.com/sot/agasc/blob/270f862458c10aaf5d1768e9711eb51b7f6fe083/create_mini_agasc_h5.py

From https://github.com/sot/agasc/pull/25/files.

Or am I still misunderstanding?

malgosias replied:
Ah, OK, my mistake. I will use the version with a command line input.

taldcroft commented on test_agasc1p7.py:
```diff
@@ -0,0 +1,47 @@
+
+import numpy as np
+import tables
```
This test script should pass on Py2 and Py3, so also `import tables3_api`.  See https://github.com/sot/tables3_api/blob/master/README.md

taldcroft commented:
>  I consider this a no-impact bug and think we should just open an issue to fix this for the next AGASC. Is there anything else we should confirm?

Agreed.  I can't think of anything else.

jeanconn commented:
Should probably independently decide if we want to do anything with the 154534513 star in the h5 version or just leave it alone.

malgosias commented:
I implemented the requested updates and added a check of all fits file comment cards, let me know if this is what you had in mind. The h5 tests still fail because of star 154534513. Should I filter on this AGASC_ID and ignore this star entry for the tests? Or will the duplicate entry be removed from the fits files?

jeanconn commented:
I know @malgosias is the one who needs to undestand that comment but what does "and then have separate handling for showing that one star is also ok"?  By "that one star" do you mean the duplicate star?  And what is "ok" in this context?  Or do you want a test showing that fetching any *other* single star works?

taldcroft commented:
@malgosias - All the tests that are documented in the comments need to pass.  The implementation of how to make this so is your challenge.  One possibility - Instead of using an ID map which will fail for the one case of a duplicate ID, could you sort by ID?  That will ensure that no entries are lost.

jeanconn commented:
@malgosias , regarding "Or will the duplicate entry be removed from the fits files?", my plan is to leave the duplicate in the fits files for this release.

taldcroft commented:
@malgosias - any progress on this?

malgosias commented:
With this version I use Table.read instead of tables.openFile for easier sorting (no id maps anymore).  This breaks on my computer because of memory but works on kadi. I tried using ``tables.openFile`` and then ``readSorted`` but the jupyter h5 file has no indexing.

I sort with respect to AGASC_ID and RA so that the two entries for the duplicate star are in the same order in all tables. The tests pass with two caveats:
- Also, it seems like ~1% of stars with a change in MAG_ACA have virtually no change in MAG_ACA_ERR. I added an ``if`` statement to catch this.
- RSV1 column: it seems that there may be an issue with creating the agasc1p7 h5 file from the fits files which I don't understand. The RSV1 column from fits files does not match the RSV1 column in the original h5 file from the jupyter notebook and ``assert np.all(c_fits == c_jupyter)`` and ``assert np.all(c_fits[ok] == c_1p6[ok])`` both fail on this column.

jeanconn commented:
> Also, it seems like ~1% of stars with a change in MAG_ACA have virtually no change in MAG_ACA_ERR. I added an if statement to catch this.

I don't understand.  Is this a error I have introduced when I made the fits files, an error in the source h5 file I had from you, or not an error at all?

malgosias commented:
> > Also, it seems like ~1% of stars with a change in MAG_ACA have virtually no change in MAG_ACA_ERR. I added an if statement to catch this.
> 
> I don't understand. Is this a error I have introduced when I made the fits files, an error in the source h5 file I had from you, or not an error at all?

Cell 87 in
http://nbviewer.jupyter.org/url/cxc.harvard.edu/mta/ASPECT/ipynb/agasc_apass_recalibration.ipynb shows that there will be stars for which the new MAG_ACA_ERR is very close (or the same) as V1.6 MAG_ACA_ERR. I will run this notebook again to confirm that this is the case and count the number of stars for which MAG_ACA_ERR does not change.

malgosias commented:
> > > Also, it seems like ~1% of stars with a change in MAG_ACA have virtually no change in MAG_ACA_ERR. I added an if statement to catch this.
> > 
> > 
> > I don't understand. Is this a error I have introduced when I made the fits files, an error in the source h5 file I had from you, or not an error at all?
> 
> Cell 87 in
> http://nbviewer.jupyter.org/url/cxc.harvard.edu/mta/ASPECT/ipynb/agasc_apass_recalibration.ipynb shows that there will be stars for which the new MAG_ACA_ERR is very close (or the same) as V1.6 MAG_ACA_ERR. I will run this notebook again to confirm that this is the case and count the number of stars for which MAG_ACA_ERR does not change.

I ran the jupyter notebook and confirmed that there are 170132 stars (in agreement with the number from my test script) for which mag_aca_err change is too small to be reflected in the h5 file where MAG_ACA_ERR column is a small integer. So for example if 1p6 MAG_ACA_ERR is 3 and new MAG_ACA_ERR is 3.3 then the h5 file will still have 3 in this column.

malgosias commented:
All implemented tests pass.

taldcroft commented:
Woohoo!
